.github/workflows/linux.yml

```diff
@@ -38,3 +38,52 @@ jobs:
         env:
           MAX_WARNINGS: ${{ matrix.conf.max_warnings }}
         run:  ./scripts/count-warnings.py build.log
+
+  build_linux_release_dynamic:
+    name: Release build (dynamic)
+    runs-on: ubuntu-16.04
+    steps:
+      - uses: actions/checkout@v1
+      - run:  sudo apt-get update
+      - name: Install C++ compiler and libraries
+        run:  sudo apt-get install -y tree $(./scripts/list-build-dependencies.sh -m apt -c gcc)
+      - name: Log environment
+        run:  ./scripts/log-env.sh
+      - name: Build
+        run: |
+          set -x
+          export VERSION=$(git describe --abbrev=4)
+          sed -i "s/AC_INIT(dosbox,git)/AC_INIT(dosbox,$VERSION)/" configure.ac
+          ./autogen.sh
+          ./configure CXXFLAGS="-O3 -DNDEBUG"
+          make -j "$(nproc)"
+          strip src/dosbox
+          # Inject variable into the surrounding env in workflow,
+          # otherwise it won't be usable in the following steps.
+          echo ::set-env name=VERSION::$VERSION
+      - name: Package
+        run: |
+          set -x
+          # Prepare content
+          install -DT        src/dosbox           dest/dosbox
+          install -DT -m 644 docs/README.template dest/README
+          install -DT -m 644 COPYING              dest/COPYING
+          install -DT -m 644 README               dest/doc/manual.txt
+          install -DT -m 644 docs/README.video    dest/doc/video.txt
+          install -DT -m 644 docs/dosbox.1        dest/man/dosbox.1
+          # Fill README template file
+          sed -i "s|%GIT_COMMIT%|$GITHUB_SHA|"               dest/README
+          sed -i "s|%GIT_BRANCH%|${GITHUB_REF#refs/heads/}|" dest/README
+          sed -i "s|%GITHUB_REPO%|$GITHUB_REPOSITORY|"       dest/README
+          mv dest "dosbox-staging-$VERSION"
+          tree --si -p "dosbox-staging-$VERSION"
+          # Create tarball
+          tar -cJf "dosbox-staging-$VERSION.tar.xz" "dosbox-staging-$VERSION"
```

Comment:
For regular testers, I think we have one too many versioned levels between the zip's directory and the tar's directory. 
Extracting one and then the other produces a double-versioned tree:

```
v0.75.0-pre-196-gb7377-lin-x86_64
└── dosbox-staging-v0.75.0-pre-196-gb7377
    ├── doc
    └── man
v0.75.0-pre-197-g2bb6-lin-x86_64
└── dosbox-staging-v0.75.0-pre-197-g2bb6
    ├── doc
    └── man
v0.75.0-pre-198-ga4d5e-linux-x86_64
└── dosbox-staging-v0.75.0-pre-198-ga4d5e
    ├── doc
    └── man
v0.75.0-pre-201-ge847-linux-x86_64
└── dosbox-staging-v0.75.0-pre-201-ge847
    ├── doc
    └── man
v0.75.0-pre-202-ge913e-linux-x86_64
└── dosbox-staging-v0.75.0-pre-202-ge913e
    ├── doc
    └── man
```

The user launches the binary with this double-versioned path:
`/my/testing/v0.75.0-pre-202-ge913e-linux-x86_64/dosbox-staging-v0.75.0-pre-202-ge913e/dosbox`

If all our zip-files unpacked into a common directory, say `dosbox-staging` we could still the versioned internal directory: 

```
dosbox-staging/
├── dosbox-staging-v0.75.0-pre-196-gb7377
│   ├── doc
│   └── man
├── dosbox-staging-v0.75.0-pre-197-g2bb6
│   ├── doc
│   └── man
├── dosbox-staging-v0.75.0-pre-198-ga4d5e
│   ├── doc
│   └── man
├── dosbox-staging-v0.75.0-pre-201-ge847
│   ├── doc
│   └── man
└── dosbox-staging-v0.75.0-pre-202-ge913e
    ├── doc
    └── man
```
The user launches the binary with a single-versioned path:
`/my/testing/dosbox-staging/dosbox-staging-v0.75.0-pre-202-ge913e/dosbox`

Another option is we version the xz name, which extracts versioned sub-files:
![2019-12-05_08-03](https://user-images.githubusercontent.com/1557255/70252181-cab84180-1735-11ea-9450-f2bc2a5c7152.png)

The user launches the binary with a single-versioned path:
`/my/testing/dosbox-staging/dosbox-staging-v0.75.0-pre-202-ge913e`

In this scenario, the version info is now directly handed to the user by their launch line, instead of having to go into the README (so we could probably do away with that; or perhaps use README to hold pretty-printed `git log` content, similar to the linux kernel's changelog).

I think the latter two options are robust enough that the user could extract their .xz into /usr/local or their home directory, and either uses versioned file/directory names so nothing is clobbered.

Reply:
I started versioning zip files to prevent the browser from appending `(1)`, `(2)` and so on to the download name, but yeah - having these directories versioned twice is not good. Appending version to the binary seems like ok solution… but all other files in the snapshot are versioned as well (especially generated README file). I think option 2 might be the way to go.

Reply:
> versioning zip files to prevent the browser from appending (1), (2)

Makes sense; we definitely want the zip-filename to be versioned, but the directory inside the zip to be generic, like `dosbox-staging`.  Something like this in the YAML maybe?
``` yaml
        with:
          name: ${{ env.VERSION }}-linux-x86_64
          path: dosbox-staging
```

> but all other files in the snapshot are versioned as well (especially generated README file). I think option 2 might be the way to go.

Good point; yes - option 2 looks to be nicest balance.  

Regarding GitHub's zipper stripping permissions (:man_facepalming: ): this is brutal, given `zip` under Linux _does_ store permissions.. so GitHub is going out of their way to do the wrong thing.  Argh!

Someone had [opened an issue](https://github.com/actions/upload-artifact/issues/10) but then closed it without resolution.  So we have a new opportunity to flag this to GitHub's developers.

I wonder if the action is smart enough to avoid zipping a zip file?  Ideally we could build our own (sanely-created) zip file that retains permissions, and provided it to the action:
``` yaml
        with:
          name: ${{ env.VERSION }}-linux-x86_64
          path: dosbox-staging-${{ env.VERSION }}-linux-x86_64.zip
```
If the action still zips the zip, then we have another avenue to create an issue, and this one will be harder for them to dismission.

Reply:
Tried double-zip, and… GitHub Actions surprised me [again](https://github.com/dreamer/dosbox-staging/runs/335693970) - I can't say if double zip was avoided, artifact was "uploaded", but in the result job can't finish, so artifact cannot be downloaded. Going back to zipping a tarball :( [edit] several minutes later it really finished and it's a zip within a zip :(

Decided to name the zip `dosbox-staging-linux-x86_64` - added `linux` to make it a bit clearer to the user, and `x86_64` to prevent confusion for users who expect to run dosbox-staging on their Raspberry Pis or Pinebooks (btw, I found a way to build ARMv7 without setting up cross-compilation!)

Reply:
OK, with fixed zip name and versioned tarball I think we found a middle ground - I can click download, open zip (instead of saving it) and decide if I want to unpack directory (and keep all snapshots in one subdirectory) or tarball. Out of all mediocre options we have, this seems to be the most flexible.

This issue was reported for this action also as actions/upload-artifact#20 and actions/upload-artifact#37.

GitHub Actions are in Beta still… but I hope they'll fix it soon-ish.

Reply:
Cool - curious how you're building for ARM!

I've opened issues 38 and 39 over at https://github.com/actions/upload-artifact/issues